### ISSUE TITLE
[FEATURE] neuron Phase 9 — group-as-node graph hidden layer foundations (architecture 구성 계획 B 진입)

### ISSUE THREAD

**Issue by juhy0987** [EinSofINTEREST/GraphLM#59]
## 배경

Phase 8 (PR #56) 으로 structural axis foundations (Growable*) 검증 완료. 다음 step 은 사용자 vision (GraphLM 원래 의도) 의 **hidden layer 자체를 graph 구조로** 구현 — Notion 의 [Phase 진행의 아키텍처 구성 계획](https://www.notion.so/36ce8b70b7aa818cbf1fe71687b449b8) 에 따라 **B (group-as-node) 먼저** 진입.

## 핵심 설계 — group-as-node graph

```
hidden_dim H 채널을 k 개씩 묶어 G = H/k 그룹 → G 개 node
within-group: dense (k × k)
between-group: 학습된 adjacency 가 결정하는 block-sparse routing
```

- node = k 채널 묶음 (1 그룹)
- edge weight = group 간 block weight (k × k block 단위)
- adjacency: 학습 가능 continuous soft mask (function-preserving init = 모두 1)
- 그룹 크기 k = head_dim 으로 시작 (Phase 1~7 의 module-level graph 와 단위 일치, 표준 64)

## 검증할 가설

1. **function preservation** — adjacency 모두 1 init 시 standard `nn.Linear` 와 forward 동일?
2. **gradient flow** — adjacency 가 학습 가능 (gradient 흐름) ?
3. **block-sparse 효과** — 일부 edge 를 0 으로 강제 (mask) 시 forward 가 expected 대로 sparse routing?
4. **inductive bias** — 그룹 단위 정보 흐름 vs dense Linear, 작은 task 에서 비교 가능?

### 신규 모듈 — `src/graphlm/neuron/graph_group.py`

- **`GroupGraphLinear(in_features, out_features, group_size, adj_init="full"|"identity")`**
  - in_features / group_size = `n_groups_in`, out_features / group_size = `n_groups_out`
  - 학습 가능 weight: shape `(n_groups_out, n_groups_in, group_size, group_size)` — group block weights
  - 학습 가능 adjacency: shape `(n_groups_out, n_groups_in)` — 그룹 간 routing scalar (continuous, init 모두 1)
  - forward: `y[go] = Σ_gi adj[go, gi] · (x[gi] @ W[go, gi])` (block matmul)
  - 검증: `in_features`, `out_features` 각각 `group_size` 의 배수
- **헬퍼**: `freeze_adjacency`, `sparsify_adjacency(threshold)` — Phase 10+ 후속용 stub

### 테스트 — `tests/neuron/test_graph_group.py`

- shape after init (`weight` 4D + `adj` 2D)
- adjacency = all-ones + identity weight init → standard Linear 와 forward 동치 (function preservation)
- gradient: weight, adjacency 둘 다 grad 흐름 확인
- in/out features 가 group_size 의 배수 아니면 ValueError
- `sparsify_adjacency(threshold)` 후 적용된 mask shape / 값 검증

### 데모 노트북 — `notebooks/02-function-level/08-phase9-group-graph-foundations.ipynb`

(노트북 prefix `08` — 02-function-level 의 8번째 = Phase 9 의 첫 노트북)

- **GroupGraphMLPLM** — Phase 8 의 `GrowableMLPLM` 의 graph 버전 (fc1, fc2 → GroupGraphLinear)
- 학습 1500 step, 비교:
  - Plain `nn.Linear` (baseline)
  - `GroupGraphLinear` (adj=full, gradient flow 확인)
  - `GroupGraphLinear` (adj=일부 sparse — manual mask 로 강제, inductive bias 효과 확인)
- 시각화: 학습된 adjacency heatmap (G × G), loss curve 비교

## 완료 조건

- [ ] `GroupGraphLinear` 구현 + function preservation 검증
- [ ] adjacency 학습 가능성 (gradient) 검증
- [ ] 신규 테스트 ≥ 5건 통과
- [ ] 데모 노트북 실행 + verdict
- [ ] `make lint` / `make test` 통과
- [ ] PR 머지

## Phase 10+ 후보

- Phase 10: **A (channel-as-node) foundations** — fine-grained sparse linear
- Phase 11: **hybrid (A + B 계층적)** 구현 — 그룹 graph 안에 채널 graph nest
- Phase 12: Transformer 통합 (Q/K/V/O 모두 GroupGraphLinear, RMSNorm 대체)
- Phase 13: 학습된 routing (DARTS / L0 / Gumbel) 로 sparsification

## 참고

- Notion 아키텍처 구성 계획: https://www.notion.so/36ce8b70b7aa818cbf1fe71687b449b8
- ML 용어집: https://www.notion.so/36ce8b70b7aa812298bbe1388e61b753
- 참조 논문: Mixture of Experts (Switch Transformer), Multi-head Attention, DeepSeek block-sparse, DARTS
